File: haskell/lib/Kan/Yoneda.hs
module Yoneda where

newtype Yoneda f a = Yoneda { runYoneda :: forall r. (a -> r) -> f r }

liftYoneda :: f a -> Yoneda f a
liftYoneda fa = Yoneda (\k -> fmap k fa)

lowerYoneda :: Yoneda f a -> f a
lowerYoneda (Yoneda m) = m id

contramap :: (b -> a) -> Yoneda f a -> Yoneda f b
contramap g (Yoneda m) = Yoneda (\k -> m (k . g))

File: rust/src/kan/yoneda.rs
mod yoneda {
    pub struct Yoneda<F, A>(Box<dyn Fn(Box<dyn Fn(A) -> R>) -> F>);

    impl<F, A> Yoneda<F, A> {
        pub fn run_yoneda<R>(&self, k: Box<dyn Fn(A) -> R>) -> F {
            (self.0)(k)
        }
    }

    pub fn lift_yoneda<F, A>(fa: F) -> Yoneda<F, A>
    where
        F: Clone,
    {
        Yoneda(Box::new(move |k| {
            fa.clone().contramap(k)
        }))
    }

    pub fn lower_yoneda<F, A>(m: Yoneda<F, A>) -> F
    where
        F: Clone,
    {
        m.run_yoneda(Box::new(|x| x.clone()))
    }

    pub fn contramap<F, A, B>(g: Box<dyn Fn(B) -> A>, m: Yoneda<F, A>) -> Yoneda<F, B> {
        Yoneda(Box::new(move |k| m.run_yoneda(Box::new(move |x| k(g(x))))))
    }
}